### ISSUE TITLE
Optimization algorithms

### ISSUE THREAD

**Issue by xuanyuw** [xuanyuw/xuanyuw.github.io#28]
https://ruder.io/optimizing-gradient-descent/index.html

## Momentum

- SGD + fraction term
- accelerate SGD in the relevant direction
- faster convergence and reduced oscillation

## Nesterov accelerated gradient (NAG)

- calculate gradient w.r.t. the approximate future position of params
- prevent from going too fast and results in increased responsiveness
- increased the perf. of RNN

## Adagrad

- adapts the learning rate to the parameters
  - low lr for param associated with frequently occurring features, larger lr for infrequent features.
- improved the robustness of SGD
- uses a different learning rate for every parameter θi at every time step t
- accumulates all past square gradients
- lr shrink and eventually become infinitesimally small

## Adadelta

- only accumulate a window of past gradients

## RMSprop

- similar to adadelta
- suggest γ = 0.9, η = 0.001

## Adam

- computes adaptive learning rates for each parameter
- storing an exponentially decaying average of past squared gradients like Adadelta and RMSprop
- also keeps an exponentially decaying average of past gradients
  mt, similar to momentum
- estimates the first moment (the mean) and the second moment (the uncentered variance) of the gradients respectively
- works well in practice

## Nadam

- NAG + Adam
- a more accurate step in the gradient direction by updating the parameters with the momentum step before computing the gradient.

## AdamW

- Use weight decay instead of L2 norm in Adam
- weight decay and learning rates can be updated separately
- improves generalization

## AMSGrad

- uses the maximum of past squared gradients rather than the exponential average to update the parameters
- outperforms adam in scenarios when adam fails to converge to an optimal solution.
- seems like it works not so great in practice
